File: 802-找到最终的安全状态/solution.cpp
#include<iostream>
#include<vector>
#include<unordered_set>
using namespace std;

class Solution
{
public:
    vector<int> eventualSafeNodes(vector<vector<int>> &graph){
        unordered_set<int> cycle;
        unordered_set<int> acycle;
        vector<int> ret;
        for (int i = 0; i < graph.size(); i++){
            if(cycle.find(i) != cycle.end()){
                continue;
            }
            if(acycle.find(i) != acycle.end()){
                ret.push_back(i);
                continue;
            }
            unordered_set<int> path;
            unordered_set<int> visited;
            bool flag = this->cycle(graph, i, path, visited);
            if(flag){
                handler(path, cycle);
            }else{
                ret.push_back(i);
                handler(visited, acycle);
            }
        }
        return ret;
    }

    void handler(unordered_set<int> &source, unordered_set<int> &target){
        for (auto iter = source.begin(); iter != source.end(); iter++){
            target.insert(*iter);
        }   
    }

    bool cycle(vector<vector<int>> &graph, int v, unordered_set<int> &path, unordered_set<int> &visited){
        visited.insert(v);
        path.insert(v);
        for (int i = 0; i < graph[v].size(); i++){
            if(path.find(graph[v][i]) != path.end()){
                return true;
            }
            if(cycle(graph, graph[v][i], path, visited)){
                return true;
            }
        }
        path.erase(v);
        return false;
    }
};

File: 802-找到最终的安全状态/solution.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        cycle, acycle = set(), set()
        ret = []
        for i in range(len(graph)):
            if i in cycle:
                continue
            if i in acycle:
                ret.append(i)
                continue
            visited, path = set(), set()
            flag = self.cycle(graph, i, visited, path)
            if flag:
                cycle = cycle | path
                acycle = acycle | (visited - path)
            else:
                acycle = acycle | visited
                ret.append(i)
        return ret

    def cycle(self, graph, v, visited, path):
        visited.add(v)
        path.add(v)
        for nbs in graph[v]:
            if nbs in path:
                return True
            if self.cycle(graph, nbs, visited, path):
                return True
        path.remove(v)
        return False
